### PR TITLE
scaffolder: `Secret` extension for use for hiding secrets securely

### LINKED ISSUE #14428: 🐛 Bug Report: Scaffolder documentation recommends insecure approach to mask secrets
### 📜 Description

The documentation at https://backstage.io/docs/features/software-templates/writing-templates#hide-or-mask-sensitive-data-on-review-step suggests usage of `ui:widget: password` to hide input. This is a pretty big footgun, since these values get stored in the tasks execution without being treated as secrets. They can therefore be viewed by other users.

### 👍 Expected behavior

Either our implementation needs to handle these values as secrets, or we need to recommend against the usage of `ui:widget: password`, or something else.

### 👎 Actual Behavior with Screenshots

We recommend an insecure approach that leaks secrets.

### 👟 Reproduction steps

1. Create a template with a field that uses `ui:widget: password`
2. Execute the template by filling out the form
3. Go to the task list and select the task run from (2), and click "Start Over"
4. Edit the DOM properties of the input to remove `type="password"` to reveal the secret that you entered in (2)

### 📃 Provide the context for the Bug.

https://github.com/backstage/backstage/pull/11154#issuecomment-1302055861

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

### PR DESCRIPTION
**Pull request by benjdlambert**
Fixes #14428 

We currently don't have a good way to fix this, so wanted to build a nice `Secret` field extension for people to use pretty easily.

There's a little bit of an issue here around having these things defined in the `parameters` schema, which means that it needs to be in the `formData`, but of course we don't want secrets sent to the backend, so there's a little bit of a hack/workaround to write _something_  to the `formData` so that the `ajv` validator from `rjsf` is happy in case there is a `required` array which selects the secret property.

Was thinking that we could get around this by implementing our own validator that doesn't care about these fields providing that they have been set in the `secrets` context instead, but that's not going to change how this works, so happy to go this way for now and iterate on this and remove the hacky parts later down the line.

Thoughts?

### CODE REVIEW

**Comment by JohnHillegass** [comment]
hey @benjdlambert, anything you need help with on this one to push it across the line? Would love to make a first contribution and seemed like a good opportunity :)

File: docs/features/software-templates/writing-templates.md [hunk 1]
```diff
@@ -231,8 +231,53 @@ spec:
             inputType: tel
 ```
 
+### Using Secrets
+
+You may want to mark thinks as secret and make sure that these values are protected and not available through REST endpoints. You can do this by using the built in `ui:field: Secret`.
```
Comment (freben) [inline docs/features/software-templates/writing-templates.md:236]:
```suggestion
You may want to mark things as secret and make sure that these values are protected and not available through REST endpoints. You can do this by using the built in `ui:field: Secret`.
```

File: docs/features/software-templates/writing-templates.md [hunk 2]
```diff
@@ -231,8 +231,53 @@ spec:
...
+
+You can define this property as any normal parameter, however the consumption of this parameter will not be available through `${{ parameters.myKey }}` you will need to use `${{ secrets.myKey }}` instead in the `template.yaml`.
```
Comment (freben) [inline docs/features/software-templates/writing-templates.md:238]:
```suggestion
You can define this property as any normal parameter, however the consumption of this parameter will not be available through `${{ parameters.myKey }}`. You will instead need to use `${{ secrets.myKey }}` in your `template.yaml`.
```

File: plugins/scaffolder/api-report.md [hunk 3]
```diff
@@ -581,6 +582,9 @@ export type ScaffolderTaskStatus = ScaffolderTaskStatus_2;
 // @public @deprecated (undocumented)
 export type ScaffolderUseTemplateSecrets = ScaffolderUseTemplateSecrets_2;
 
+// @public (undocumented)
+export const Secret: (props: ScaffolderRJSFFieldProps) => React_2.JSX.Element;
```
Comment (freben) [inline plugins/scaffolder/api-report.md:586]: i see that other controls also follow this simple pattern, but with this single-word control it looks like it'd been nice if they'd been grouped under some namespace or at least been named `FooFormControl` or whatnot

File: plugins/scaffolder/src/components/fields/Secret/Secret.test.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        />
+        <SecretsComponent />
+      </SecretsContextProvider>,
+    );
+
+    const secretInput = getByLabelText('secret');
+    const submitButton = getByRole('button');
+
+    await act(async () => {
+      fireEvent.change(secretInput, { target: { value: mockSecret } });
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/Secret/Secret.test.tsx:60]: matter of taste: we should be using userEvent more eh, it's nicely compounding things and deals with `act` under the hood etc

File: plugins/scaffolder/src/components/fields/Secret/Secret.test.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    return (
+      <div data-testid="current-secrets">{JSON.stringify({ secrets })}</div>
+    );
+  };
+
+  it('should set the current form value as a mask for the value entered', async () => {
+    const mockSecret = 'backstage';
+    const onSubmit = jest.fn();
+
+    const { getByLabelText, getByRole } = await renderInTestApp(
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/Secret/Secret.test.tsx:39]: matter of taste: `screen` is real nice imo

**Review by freben — APPROVED** [review]

File: docs/features/software-templates/writing-templates.md (hunk 1 above)
Comment (benjdlambert) [reply docs/features/software-templates/writing-templates.md:236]: 🤔

File: plugins/scaffolder/api-report.md (hunk 3 above)
Comment (benjdlambert) [reply plugins/scaffolder/api-report.md:586]: Yeah, I wonder if we can avoid exporting this raw component actually, and just export the field extension itself.

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx [hunk 6]
```diff
@@ -111,6 +113,31 @@ export const Stepper = (stepperProps: StepperProps) => {
...
+        validator.rawValidation(schema, { ...formData, ...secrets }),
+      validateFormData: (formData, schema, customFormats, transformErrors) =>
+        validator.validateFormData(
+          { ...formData, ...secrets },
+          schema,
+          customFormats,
+          transformErrors,
+        ),
+      // @deprecated
+      toErrorList: validator.toErrorList,
```
Comment (freben) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:135]: just note that this doesn't preserve `this` === `validator` - maybe not a problem in practice

File: plugins/scaffolder/src/components/fields/SecretInput/SecretInput.test.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  SecretsContextProvider,
+  useTemplateSecrets,
+} from '@backstage/plugin-scaffolder-react';
+import { SecretInput } from './SecretInput';
+import { renderInTestApp } from '@backstage/test-utils';
+import { Form } from '@backstage/plugin-scaffolder-react/alpha';
+import validator from '@rjsf/validator-ajv8';
+import { fireEvent, act } from '@testing-library/react';
+
+describe('<Secret/>', () => {
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/SecretInput/SecretInput.test.tsx:27]:
```suggestion
describe('<SecretInput />', () => {
```

**Review by freben — APPROVED** [review]

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx (hunk 6 above)
Comment (benjdlambert) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:135]: Yeah it's not called at all, it's just there to make typescript happy.

**Review by Rugvip — APPROVED** [review]
Nice! 👍 🎉

**Pull request merged**